File: dsa-problems/NumberTheory/solutions/javascript/NUM-015-crt-existence-value.js
const readline = require("readline");

function extendedGCD(a, b) {
  if (b === 0n) return [a, 1n, 0n];
  const [g, x1, y1] = extendedGCD(b, a % b);
  const x = y1;
  const y = x1 - (a / b) * y1;
  return [g, x, y];
}

function crtSolve(a, m) {
  let curA = 0n;
  let curM = 1n;
  
  for (let i = 0; i < a.length; i++) {
    const A = BigInt(a[i]);
    const M = BigInt(m[i]);
    
    let rhs = (A - curA) % M;
    if (rhs < 0n) rhs += M;
    
    const [g, x, y] = extendedGCD(curM, M);
    
    if (rhs % g !== 0n) return null;
    
    const mod = M / g;
    let k = (rhs / g) * (x % mod + mod) % mod; // x is inverse of curM/g
    
    const newM = curM * (M / g);
    curA = (curA + k * curM) % newM;
    curM = newM;
  }
  
  return curA;
}

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let data = [];
rl.on("line", (line) => {
  const parts = line.trim().split(/\s+/);
  for (let i = 0; i < parts.length; i++) {
    if (parts[i].length > 0) data.push(parts[i]);
  }
});
rl.on("close", () => {
  if (data.length === 0) return;
  let idx = 0;
  const k = parseInt(data[idx++], 10);
  const a = [];
  const m = [];
  for (let i = 0; i < k; i++) {
    a.push(parseInt(data[idx++], 10));
    m.push(parseInt(data[idx++], 10));
  }
  const res = crtSolve(a, m);
  console.log(res === null ? "NO" : res.toString());
});

File: dsa-problems/NumberTheory/solutions/python/NUM-015-crt-existence-value.py
import sys

def extended_gcd(a, b):
    if b == 0:
        return a, 1, 0
    g, x1, y1 = extended_gcd(b, a % b)
    x = y1
    y = x1 - (a // b) * y1
    return g, x, y

def crt_solve(a, m):
    cur_a = 0
    cur_m = 1
    
    for i in range(len(a)):
        A = a[i]
        M = m[i]
        
        # k * cur_m = A - cur_a (mod M)
        rhs = (A - cur_a) % M
        
        g, x, y = extended_gcd(cur_m, M)
        
        if rhs % g != 0:
            return None
            
        inv = x % (M // g)
        k = (rhs // g) * inv % (M // g)
        
        # New state
        # cur_a = cur_a + k * cur_m
        # cur_m = lcm(cur_m, M)
        
        new_m = cur_m * (M // g)
        cur_a = (cur_a + k * cur_m) % new_m
        cur_m = new_m
        
    return cur_a

def main():
    input = sys.stdin.read
    data = input().split()
    if not data:
        return
    iterator = iter(data)
    try:
        k = int(next(iterator))
        a = []
        m = []
        for _ in range(k):
            a.append(int(next(iterator)))
            m.append(int(next(iterator)))
            
        res = crt_solve(a, m)
        print("NO" if res is None else res)
    except StopIteration:
        pass

if __name__ == "__main__":
    main()
